cli/command/stack/kubernetes/services.go

```diff
@@ -2,43 +2,130 @@ package kubernetes
 
 import (
 	"fmt"
+	"sort"
+	"strings"
 
 	"github.com/docker/cli/cli/command/formatter"
 	"github.com/docker/cli/cli/command/stack/options"
 	"github.com/docker/cli/kubernetes/labels"
+	"github.com/docker/docker/api/types/filters"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// RunServices is the kubernetes implementation of docker stack services
-func RunServices(dockerCli *KubeCli, opts options.Services) error {
-	// Initialize clients
+var supportedServicesFilters = map[string]bool{
+	"mode":  true,
+	"name":  true,
+	"label": true,
+}
+
+func generateValuedLabelsSelector(valuedLabels map[string][]string) []string {
+	var result []string
+	for k, v := range valuedLabels {
+		if len(v) == 1 {
+			result = append(result, fmt.Sprintf("%s=%s", k, v[0]))
+		} else {
+			sort.Strings(v)
+			result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
+		}
+	}
+	return result
+}
+
+func parseLabelFilters(rawFilters []string) ([]string, map[string][]string) {
+	var presentLabels []string
+	valuedLabels := map[string][]string{}
+	for _, rawLabel := range rawFilters {
+		equalIndex := strings.Index(rawLabel, "=")
+		if equalIndex == -1 {
+			presentLabels = append(presentLabels, rawLabel)
+		} else {
+			key := rawLabel[:equalIndex]
+			val := rawLabel[equalIndex+1:]
+			valuedLabels[key] = append(valuedLabels[key], val)
+		}
+	}
+	return presentLabels, valuedLabels
+}
+
+func generateLabelSelector(f filters.Args, stackName string) string {
+	names := f.Get("name")
+	sort.Strings(names)
+
+	for _, n := range names {
+		if strings.HasPrefix(n, stackName+"_") {
+			// also accepts with unprefixed service name (for compat with existing swarm scripts where service names are prefixed by stack names)
```

Comment:
Same question regarding backward-combat: if this is only for stacks created with the experimental feature, we may want to consider breaking backward compatibility (and just document)

Reply:
My understanding is that Kubernetes, unlike Swarm, does not prefix its service names with the name of the stack, therefore as a convenience to match the likeness of both outputs, the docker CLI decorates the names of the Kubernetes services to add the same style of prefix.